mipengine/controller/node_landscape_aggregator.py

```diff
@@ -274,67 +276,78 @@ def get_node_specific_datasets(
-        for data_model, datasets in datasets_per_data_model.items():
-            if data_model not in aggregated_datasets:
-                aggregated_datasets[data_model] = []
-            aggregated_datasets[data_model].extend(datasets.keys())
-
-    duplicated_datasets = {
-        data_model: [
-            item for item, count in Counter(dataset_names).items() if count > 1
-        ]
-        for data_model, dataset_names in aggregated_datasets.items()
-    }
 
-    datasets_per_node_without_duplicates = {
-        node_id: {
-            data_model: {
-                dataset_name: dataset_label
-                for dataset_name, dataset_label in datasets.items()
-                if dataset_name not in duplicated_datasets[data_model]
-            }
-            for data_model, datasets in datasets_per_data_model.items()
-        }
-        for node_id, datasets_per_data_model in datasets_per_node.items()
-    }
+def data_model_registry_data_cruncing(
+    federation_info: FederationInfo,
+) -> DataModelRegistry:
+    federation_info_with_compatible_data_models = _remove_incompatible_data_models(
+        federation_info
+    )
+    datasets_locations, aggregated_datasets = _gather_all_dataset_info(
+        federation_info_with_compatible_data_models
+    )
+    data_models = _get_data_models(
+        federation_info_with_compatible_data_models, aggregated_datasets
+    )
 
-    _log_duplicated_datasets_per_node(
-        datasets_per_node, datasets_per_node_without_duplicates
+    return DataModelRegistry(
+        data_models=data_models, datasets_locations=datasets_locations
     )
 
-    return datasets_per_node_without_duplicates
 
+def _get_federation_info(
+    nodes: List[NodeInfo],
+) -> FederationInfo:
+    federation_info = {}
+
+    for node_info in nodes:
+        dmi = {}
+
+        node_socket_addr = _get_node_socket_addr(node_info)
+        datasets_per_data_model = _get_node_datasets_per_data_model(node_socket_addr)
+        if datasets_per_data_model:
+            node_socket_addr = _get_node_socket_addr(node_info)
+            for data_model, datasets in datasets_per_data_model.items():
+
+                cdes = _get_node_cdes(node_socket_addr, data_model)
+                if not cdes:
+                    continue
+                dmi[data_model] = (DatasetsInfo(values=datasets), cdes)
+        federation_info[node_info.id] = DataModelsInfo(values=dmi)
 
-def _gather_all_dataset_info(
-    datasets_per_node: Dict[str, Dict[str, Dict[str, str]]],
-) -> Tuple[Dict[str, Dict[str, str]], Dict[str, Dict[str, str]]]:
+    return FederationInfo(values=federation_info)
+
+
+def _gather_all_dataset_info(federation_info):
     """
     Args:
-        datasets_per_node: The datasets for each node available in the system
+        federation_info
     Returns:
         A tuple with:
          1. The location of each dataset.
          2. The aggregated datasets, existing in all nodes
     """
```

Comment:
instead of relying on comments to convey the meaning of a returned tuple you could define data transfer objects with named fields(variables) that are self explanatory